src/tree/block_scanner.rs

```diff
@@ -27,47 +34,77 @@ where
     pub async fn new(
         middleware: Arc<M>,
         window_size: u64,
-        current_block: u64,
+        start_block: u64,
         filter: Filter,
-    ) -> Result<Self, M::Error> {
+    ) -> WorldTreeResult<Self> {
         let chain_id = middleware.get_chainid().await?.as_u64();
         Ok(Self {
             middleware,
-            next_block: Mutex::new(current_block),
+            start_block,
             window_size,
             filter,
             chain_id,
         })
     }
 
-    /// Retrieves events matching the specified address and topics from the last synced block to the latest block, stepping by `window_size`.
-    /// Note that the logs are unsorted and should be handled accordingly.
-    pub async fn next(&self) -> Result<(usize, Vec<Log>), M::Error> {
-        let mut next_block = self.next_block.lock().await;
-        let latest_block = self.middleware.get_block_number().await?.as_u64();
-
-        let to_block = (*next_block + self.window_size).min(latest_block);
-
-        let num_blocks =
-            if let Some(num_blocks) = to_block.checked_sub(*next_block) {
-                num_blocks as usize
-            } else {
-                return Ok((0, vec![]));
+    pub fn block_stream(
+        &self,
+    ) -> impl Stream<Item: Future<Output = WorldTreeResult<Vec<Log>>> + Send> + '_
+    {
+        stream::unfold(self.start_block, move |mut next_block| async move {
+            // This is executed before the item is yielded
+            let to_block = loop {
+                match self.middleware.get_block_number().await {
```

Comment:
Previously we have seen connection closed errors occasionally. IIRC ethers does not have reconnect logic under the hood. In the case that the connection closes, would we perpetually loop here?

Reply:
Yes it's possible we perpetually loop here. How would you fix that though, a limited number of retries?

Reply:
Yes we could potentially exit after a specified number of retries. We could also pass the rpc url  to the block scanner and try to reconnect after repeated failures.